### ISSUE TITLE
[RFC] M-to-N cast primitive (for RL weight rollout & pipeline parallel)

### ISSUE THREAD

**Issue by kwen2501** [pytorch/pytorch#181061]
## TL;DR

Reinforcement-learning training loops spend a large and growing share of their wall-clock time on the **Refit** step — moving updated weights from a *trainer* cluster to a *generator* cluster that uses a different parallelism layout. We propose adding a first-class **M-to-N resharding collective** to `torch.distributed` so that frameworks like Llama-RL, NeMo-RL, veRL, and others can express "send these parameters from the trainer's sharding to the generator's sharding" as a single load-balanced, topology-aware call, instead of stitching together `all_gather` + `broadcast` or hand-rolled point-to-point routes.

A working prototype (API sketched below) lives on top of `torch.distributed._symmetric_memory` and drops refit time substantially at the scales we have benchmarked. We'd like to discuss the public API, the placement in the distributed stack, and what we need from NCCL / accelerator backends before upstreaming.

## Background: why RL refits matter

The canonical RL training loop is a continuous pipeline:

```
Generator -> Reward model -> Trainer -> Generator
```

Production RL systems typically partition this pipeline across *disjoint* GPU pools:

- **Trainer** cluster: runs the policy gradient update. Uses whatever (TP, PP, DP, FSDP, EP, …) strategy fits the training workload.
- **Generator** cluster: runs inference / rollouts. Uses a completely different parallelism strategy (different TP, usually no PP, smaller DP) because inference has different compute/memory trade-offs.

After every trainer step (or every N steps), the **Refit** phase has to broadcast the updated weights from the trainer layout into the generator layout. This is an *M-to-N data redistribution*: M trainer ranks collectively own the global weight tensor in one layout, N generator ranks need it in another layout, and neither set of ranks has a view that matches the other.

### The cost, empirically

Refit is already a first-order cost at today's model scales, and it grows with both model size and world size:

| Model            | Scale    | Refit share of iteration time |
| ---------------- | -------- | ----------------------------- |
| Qwen_30b_a3b     | 32 GPUs  | ~16.5%                        |
| DeepSeek-V3      | 256 GPUs | ~20%                          |

These are numbers from the workloads we've profiled. The trend is clear: as you scale, refit stalls the entire pipeline, and every percent of refit time is idle generator + idle trainer. Unlike training-side collectives (`all_reduce`, `reduce_scatter`), refit sits on the pipeline critical path — generators cannot begin the next rollout batch until the new weights arrive.

## Where `torch.distributed` is today

`torch.distributed` has historically focused on *symmetric* collectives — every rank plays the same role in the same group — because that matches the overwhelming majority of pre-training and fine-tuning workloads, where a single parallelism strategy is chosen up front and every rank cooperates inside it. This has been enormously successful: the existing primitives (`all_reduce`, `reduce_scatter`, `all_gather`, `all_to_all`, `broadcast`, `send`/`recv`) cover essentially all of classical distributed training.

The RL refit pattern is newer and sits slightly outside this design center. It is asymmetric along two axes at once:

1. **Two distinct rank sets** (trainer vs. generator) that are logically one workload but physically disjoint — so neither set can describe the other with a single `ProcessGroup` it owns.
2. **Two different tensor layouts** on those rank sets — so the operation is not "move a tensor" but "rearrange a tensor," which requires a joint plan across both sets.

With today's primitives, frameworks can approximate this with two common patterns, both of which work correctly but leave performance on the table:

1. **All-gather + broadcast** (e.g. NeMo-RL style).
   - Trainer ranks `all_gather` each parameter to a replicated form, then `broadcast` it to the generator cluster (typically via a bridge rank or through a group spanning both clusters).
   - **Cost:** moves roughly **2x the theoretically necessary volume**. The gather materializes the full tensor on every trainer rank before broadcast, and the broadcast then re-fans it out. Any locality that already exists between the two layouts is discarded.
   - **Memory:** requires a full-tensor staging buffer, which is prohibitive for large MoE or dense models at high TP.

2. **Hand-rolled point-to-point** (e.g. Llama-RL style).
   - Compute per-parameter the set of (trainer_rank, generator_rank, byte_range) triples and issue `isend`/`irecv` pairs directly.
   - **Cost:** minimal volume in principle, but:
     - No global awareness — every pair picks its own schedule, and hot links get oversubscribed while idle links sit unused. This shows up as poor bandwidth utilization at scale.
     - No load-balancing across redundant replicas on the generator side (common when DP > 1 on the generator).
     - Resiliency and failure recovery are ad hoc; there is no single control plane.
     - Each framework reimplements this from scratch with subtle bugs.

Neither pattern is a `torch.distributed` shortcoming so much as a consequence of the fact that the fundamental operation — **"reshard this tensor from layout A on rank set A to layout B on rank set B"** — does not yet have a primitive. We think adding it would be a natural extension of the existing collective vocabulary, not a replacement for any of it.

## Proposal

Add a new collective to `torch.distributed._symmetric_memory` that expresses M-to-N resharding directly:

```python
# Sketch - exact name/signature subject to RFC discussion
torch.distributed._symmetric_memory.mxn_cast(
    buf: torch.Tensor,
    src_local_shape: Sequence[int],
    src_mesh: DeviceMesh,
    src_placements: Sequence[Placement],
    dst_local_shape: Sequence[int],
    dst_mesh: DeviceMesh,
    dst_placements: Sequence[Placement],
    group: ProcessGroup | str,
) -> None
```

Semantics:

- `src_mesh` and `dst_mesh` may be **disjoint** (the RL refit case: trainer ranks ≠ generator ranks), **overlapping** (in-place reshard within a single cluster, e.g. switching TP dims mid-pipeline), or **identical** (degenerate case).
- A rank's role is inferred from mesh membership:
  - In `src_mesh` only → *producer*. Reads its local shard from `buf`; `buf` is left unchanged on return.
  - In `dst_mesh` only → *consumer*. On return, `buf` holds this rank's local shard of the destination layout.
  - In both → in-place dual role; `buf` is read as the source layout then overwritten with the destination layout.
- Placements follow the existing DTensor `Shard(dim)` / `Replicate()` vocabulary — so a user who already describes their trainer and generator tensors as DTensors can hand those placements straight in.

The primitive takes full responsibility for:

- **Volume minimization**: each byte is transferred at most once over the cross-cluster fabric. No intermediate full-tensor gather.
- **Load balancing**: when the destination layout has replication (typical for generator DP > 1), the primitive is free to pick which source replica feeds which destination replica to spread load evenly across NICs and links.
- **Topology awareness**: intra-node (NVLink / LSA) and inter-node (RDMA / rail) transfers are planned together so the slow tier isn't oversubscribed.

### What this enables

- **RL frameworks** (NeMo-RL, Llama-RL, veRL, torchstore, TRL, OpenRLHF, etc.) can replace their hand-rolled refit paths with one call per parameter (or one call per coalesced group of parameters).
- **DTensor** gets an efficient, topology-aware "redistribute" operator for the cross-mesh case; today `redistribute()` is restricted to the same mesh.
- **Pipeline parallelism** gets a clean primitive for mid-run topology changes (e.g. previous stage has SP=2, next stage has TP=4, thus requiring a 2-to-4 cast).
- **Elastic training**: rescaling the trainer mid-run (adding / removing nodes) needs exactly this primitive to move weights from the old layout to the new one.

## Prior art and related work

- **`torch.distributed.tensor.redistribute`**: works within a single `DeviceMesh`; does not cross disjoint meshes. Cross-mesh is exactly the RL case.
- **Megatron-LM / DeepSpeed weight conversion tools**: operate offline, through CPU / filesystem. Not viable for online RL refits.
- [`meta-pytorch/torchstore`](https://github.com/meta-pytorch/torchstore): source ranks `put` their DTensor shards into a logical store; destination ranks `get` with their own target DTensor. This is the most sophisticated public approach today. It gets the volume accounting right, preserves DTensor placement information end-to-end, and confirms that DTensor placements are the right vocabulary for describing the problem. Traffic pattern wise, it is a set of _direct_ P2P operations, thus similar to the Llama-RL style mentioned above, but with one-sided improvement. Direct P2Ps could leave some performance on the table -- if a _collective_ has **topology awareness**, it could schedule the traffic within the communicator so that the least amount of traffic flows through the slowest link, and rely on faster link for relay of **duplicated** traffic.

## Layering of this op

The proposed is to add a tensor collective at `torch.distributed` level, similar to `all_gather` etc. It can be called by user-facing layers such as torchstore, NeMo-RL or veRL, etc. In addition, by being in torch, we hope the op can be used in other cases, such as pipeline parallel where the neighboring stage dimensions do not match. 
 
cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @weifengpy

**Comment by tianyu-l** [comment]
sounds overlapped with https://github.com/meta-pytorch/torchstore

cc @LucasLLC @daniellepintz @amirafzali

**Comment by amirafzali** [comment]
Hey @kwen2501, we’ve been working on a similar problem in torchstore. It supports ideas like sharded state dict exchange, DTensor-aware resharding across different meshes, and GPU-direct or buffered RDMA weight sync. I believe some of the RL refit use case you describe is covered there in practice, like cross-layout trainer/generator weight transfer. A focus for us has been a data design that minimizes the trainer/generator stall time (like in async RL). would be happy to compare notes if that would be helpful!

**Comment by kwen2501** [comment]
Apology for my negligence, I added torchstore to the **Prior art and related work** section. 
> [meta-pytorch/torchstore](https://github.com/meta-pytorch/torchstore): source ranks put their DTensor shards into a logical store; destination ranks get with their own target DTensor. This is the most sophisticated public approach today. It gets the volume accounting right, preserves DTensor placement information end-to-end, and confirms that DTensor placements are the right vocabulary for describing the problem. Traffic pattern wise, it is a set of direct P2P operations, thus similar to the Llama-RL style mentioned above, though with one-sided improvement. Direct P2Ps could leave some performance on the table; if the tensor operator has topology awareness, it could schedule the traffic within the communicator so that the least amount of traffic flows through the slowest link, and rely on faster link for relay of duplicated traffic.

**Comment by kwen2501** [comment]
(continued) A natural reading is that torchstore has already validated the cross-mesh DTensor pattern at the *system* level and surfaced the gaps a lower-level *collective* primitive could close. A future torchstore could use such a primitive as its resharding backend — the two are complementary, not competing.

**Comment by d4l3k** [comment]
I think the primitive here is fairly reasonable and something that makes sense in SymmetricMemory. As for it being a top level function in torch.distributed, that's an open question. Historically those have all done through the PG API but supporting symmetric memory operations there opens up this for a lot more flexibility to support new operations. 

My main concern for promotion to torch.distributed is that we should support at least two backends and the specific API. Is it too tied to DeviceMesh? Should we have it be a lower level function and then have a separate helper in DeviceMesh to handle this operation?

**Comment by d4l3k** [comment]
cc @ngimel

**Comment by kwen2501** [comment]
@d4l3k Thanks. Indeed, if it has a symmetric input requirement, it should be in SymmetricMemory. That works.
The low-level cpp API will not have DeviceMesh dependency. It is just the python surface using that for the convenience of callers.
